compiler/src/iree/compiler/InputConversion/TOSA/TosaToLinalgExt.cpp

```diff
@@ -0,0 +1,178 @@
+// LinalgExt version is not batched therefore we materialize the batch index
+// for each update.
+class ScatterConversion : public OpRewritePattern<tosa::ScatterOp> {
+ public:
+  using OpRewritePattern<tosa::ScatterOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::ScatterOp op,
+                                PatternRewriter &rewriter) const final {
+    auto values = op.getValuesIn();
+    auto indices = op.getIndices();
+    auto updates = op.getInput();
+    auto valuesTy = values.getType().dyn_cast<RankedTensorType>();
+    auto indicesTy = indices.getType().dyn_cast<RankedTensorType>();
+    auto updatesTy = updates.getType().dyn_cast<RankedTensorType>();
+    ImplicitLocOpBuilder builder(op.getLoc(), rewriter);
+
+    if (!valuesTy || !indicesTy || !updatesTy)
+      return rewriter.notifyMatchFailure(op,
+                                         "tosa.gather has unknown input rank");
+
+    // TOSA's scatter does not include a index dimension, instead it implicitly
+    // supporteds an index depth of one. We materialize that implicit index of
+    // one as follows: [batch, updates] -> [batch, updates, index_depth=1] With
+    // a indexing map of [[0], [1, 2]].
+    llvm::SmallVector<int64_t> expandIndShape{indicesTy.getDimSize(0),
+                                              indicesTy.getDimSize(1), 1};
+    SmallVector<ReassociationExprs, 4> expandIndMap;
+    expandIndMap.push_back({
+        builder.getAffineDimExpr(0),
+    });
+    expandIndMap.push_back({
+        builder.getAffineDimExpr(1),
+        builder.getAffineDimExpr(2),
+    });
+
+    indices = builder.create<tensor::ExpandShapeOp>(
+        indicesTy.clone(expandIndShape), indices, expandIndMap);
+    indicesTy = indices.getType().dyn_cast<RankedTensorType>();
+
+    // Materialize the batch indice as LinalgExt scatter is not batched.
+    {
+      llvm::SmallVector<Value> dynDims;
+      for (int i = 0, s = indicesTy.getRank(); i < s; ++i)
+        if (indicesTy.isDynamicDim(i))
+          dynDims.push_back(builder.create<tensor::DimOp>(indices, i));
+
+      Value empty = builder.create<tensor::EmptyOp>(
+          indicesTy.getShape(), indicesTy.getElementType(), dynDims);
+
+      Value batchIdx = nullptr;
+
+      if (indicesTy.getDimSize(0) == 1) {
+        Value zero = builder.create<arith::ConstantOp>(
+            rewriter.getZeroAttr(indicesTy.getElementType()));
+        batchIdx = builder.create<linalg::FillOp>(zero, empty).getResult(0);
+      } else {
+        SmallVector<utils::IteratorType> iterators(
+            indicesTy.getRank(), utils::IteratorType::parallel);
+        SmallVector<AffineMap, 3> indexingMaps(
+            2, builder.getMultiDimIdentityMap(indicesTy.getRank()));
+        batchIdx = builder
+                       .create<linalg::GenericOp>(
+                           indicesTy, indices, empty, indexingMaps, iterators,
+                           [&](OpBuilder &nestedBuilder, Location nestedLoc,
```

Comment:
Nit Can we assign this lambda to a variable and use here? (the amount of whitespace make this more difficult to read).

Reply:
Done.